target_bigquery.py

```diff
@@ -265,57 +276,276 @@ def persist_lines_stream(project_id, dataset_id, lines=None, validate_records=Tr
+                )
+            )
         else:
-            logging.error('Errors:', errors[table], sep=" ")
+            logger.error("Errors:", errors[table], sep=" ")
 
     return state
 
+
+def persist_lines_hybrid(project_id, dataset_id, lines=None, validate_records=True, location=None):
+    state = None
+    schemas = {}
+    key_properties = {}
+    tables = {}
+    recreated_tables = {}
+    rows = {}
+    errors = {}
+    failed_lines = []
+
+    bigquery_client = bigquery.Client(project=project_id)
+    dataset_ref = f"{project_id}.{dataset_id}"
+    dataset = bigquery.Dataset(dataset_ref)
+    if location:
+        dataset.location = location
+    bigquery_client.create_dataset(dataset, exists_ok=True)
+
+    def write_rows_to_bigquery(streams, emit_state_after_write=False):
+        nonlocal failed_lines
+        for stream in streams:
+            if rows[stream]:
+                # By using `insert_rows_json` and passing generated `row_ids` we avoid duplication
+                ids = [
+                    "-".join(str(row[val]) for val in key_properties[stream])
+                    for row in rows[stream]
+                ]
+
+                # Singer uses Decimal in the deserialised data which `insert_rows_json` can't
+                # serialise with the built in `json` class so we need to fix it
+                fixed_rows = [
+                    {k: (float(v) if isinstance(v, Decimal) else v) for (k, v) in row.items()}
+                    for row in rows[stream]
+                ]
+
+                # NOTE: as it turns out it takes BigQuery ~2 minutes to empty cache and acknowledge
+                # a new table schema, see: https://stackoverflow.com/a/25292028/21217
+                # So we allow a long retry period for recreated tables, short for incremental sync
+                max_run_time = datetime.now() + timedelta(
+                    seconds=300 if recreated_tables.get(stream) else 30
+                )
+                while max_run_time > datetime.now():
+                    # NOTE: This will fail if there are more than 10000 rows or the request size
+                    # exceeds 10MB, see: https://cloud.google.com/bigquery/quotas#streaming_inserts
+                    try:
+                        errors[stream] = bigquery_client.insert_rows_json(
+                            tables[stream], fixed_rows, row_ids=ids
+                        )
+                    except Exception as e:
+                        error_string = str(e)
+                        logger.warning(
+                            f"Error on insert_rows_json: {error_string}", extra={"stream": stream},
+                        )
+                        if (
+                            "payload size exceeds the limit" in error_string
+                            or "too many rows present" in error_string
+                        ):
+                            # Let's try inserting it in two halves
```

Comment:
If this also fails, then this might be because the lines are very long. What would we do next - break it up into even smaller chunks?

Reply:
This currently only happens in the case of lines too long – with I think module state triggering it on occasion – but I also added the too many rows case just in case.

I didn't feel like making it into a recursive function, if a table needs even smaller chunks it would probably be better to decrease it on the tap side anyway.